src/tools/imgtool/imgtool.cpp

```diff
@@ -1966,18 +1963,19 @@ imgtoolerr_t imgtool::partition::put_file(const char *newfname, const char *fork
 		newfname = basename.c_str();
 	}
 
-	charset = (imgtool_charset) (int) get_info_int(IMGTOOLINFO_INT_CHARSET);
-	if (charset != IMGTOOL_CHARSET_UTF8)
+	imgtool::charconverter *charconverter = (imgtool::charconverter *) get_info_ptr(IMGTOOLINFO_PTR_CHARCONVERTER);
+	if (charconverter)
 	{
-		/* convert to native format */
-		alloc_newfname = native_from_utf8(charset, newfname);
-		if (alloc_newfname == nullptr)
+		// convert to native format
+		try
 		{
-			err = (imgtoolerr_t)(IMGTOOLERR_BADFILENAME | IMGTOOLERR_SRC_NATIVEFILE);
-			goto done;
+			alloc_newfname = charconverter->from_utf8(newfname);
 		}
-
-		newfname = alloc_newfname;
+		catch (charconverter_exception)
+		{
+			return (imgtoolerr_t)(IMGTOOLERR_BADFILENAME | IMGTOOLERR_SRC_NATIVEFILE);
```

Comment:
Extra parentheses around the type name are clutter - C++ has function-syntax cast for enums.